In [1]:
import numpy as np
import pandas as pd
#Textblob features allows us to determine weather the input textual data has a positive, negetive, or neutral tone.
from textblob import TextBlob
import plotly.graph_objects as go
import plotly.express as xp

In [2]:
#Reading the CSV files
modi = pd.read_csv("C:/Users/ACER/OneDrive/Desktop/external project/Indian-Election-Sentiment-Analysis-main/modi_reviews.csv")
rahul = pd.read_csv('C:/Users/ACER/OneDrive/Desktop/external project/Indian-Election-Sentiment-Analysis-main/rahul_reviews.csv')

In [3]:
modi.head(5)

Unnamed: 0             User  \
0           0  advosushildixit   
1           1           jiaeur   
2           2    PVenkatGandhi   
3           3      TheNirbhay1   
4           4      ShakeChilli   

                                               Tweet  
0  @anjanaomkashyap I am seeing you as future #bj...  
1  #LokSabhaElections2019 \n23rd May 2019 will re...  
2  #LokSabhaElections2019 \n23rd May 2019 will re...  
3  PM Modi creates a new record of being the only...  
4  @abhijitmajumder Appointment of Successor! \n\...

In [4]:
#No of Comments/Messages/Rows of Modi Reviews
modi.shape

(25688, 3)

In [5]:
rahul.head(5)

Unnamed: 0             User  \
0           0     Sunnysweet16   
1           1    drnitinchaube   
2           2        mrvivek07   
3           3    JosephPravinP   
4           4  VandanaMegastar   

                                               Tweet  
0  Wonder why no academic or journalist asks INC ...  
1  Congrats for the change #australiavotes2019 an...  
2  Peopel Say “Govt Ne 70 Years Kya kiya”.\nUnse ...  
3  @ajaymaken @RahulGandhi And as a final touch, ...  
4  #LokSabhaElections2019 Anyone not having mass ...

In [6]:
#No of Comments/Messages/Rows of Rahul Reviews
rahul.shape

(14261, 3)

In [7]:
#First tweet in modi reviews
modi['Tweet'][0]

'@anjanaomkashyap I am seeing you as future #bjp spokesperson..\nGood luck. Anjana Om Modi oops Kashyap.\nJournalists like you changed the meaning of journalism. Janta maaf nai karege'

In [8]:
#Find the sentiment for the first tweet weather it is positive or negitive
#(If the value of polarity is-0.35 then the opnion of people would be negitive)
TextBlob(modi['Tweet'][0]).sentiment

Sentiment(polarity=0.35, subjectivity=0.36250000000000004)

In [9]:
#Tenth tweet in rahul reviews
rahul['Tweet'][10]

'@IYCTelangana @IncBasavakalyan @narendramodi @IYC @INCIndia @RahulGandhi @priyankagandhi @HardikPatel_ @Allavaru @keshavyadaviyc @srinivasiyc @srivatsayb @RakshaRamaiah @JM_Scindia This is called in #BJP language,\n"A Swatch &amp; Swasth Hindutva" !!! This theory not applicable to any other person, except #BJPians! \nQuestion: Demonic possession of !!!!'

In [10]:
#Find the sentiment for the tenth tweet weather it is positive or negitive
TextBlob(rahul['Tweet'][10]).sentiment

Sentiment(polarity=-0.3814697265625, subjectivity=0.375)

In [11]:
#checking the data type here as we got an error that the datatype for "Tweet" is object and we have to change it to String type 
modi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25688 entries, 0 to 25687
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  25688 non-null  int64 
 1   User        25683 non-null  object
 2   Tweet       25683 non-null  object
dtypes: int64(1), object(2)
memory usage: 602.2+ KB


In [12]:
#Changing the data type of "Tweets" to String
modi['Tweet'] = modi['Tweet'].astype(str)
rahul['Tweet'] = rahul['Tweet'].astype(str)

#This function will itarate through all the tweets and give you the review that it is positive or negitive
def find_polarity(review):
    return TextBlob(review).sentiment.polarity

In [13]:
#Making a new column to show that each Tweet is positive or negitive
modi['Polarity'] = modi['Tweet'].apply(find_polarity)
rahul['Polarity'] = rahul['Tweet'].apply(find_polarity)

# Polarity > 0= Positive, Polarity < 0= Negitive, Polarity = 0= Netural

In [14]:
#All Positive values in modi review
np.where(modi['Polarity']>0)

(array([    0,     1,     2, ..., 25679, 25681, 25684], dtype=int64),)

In [15]:
#All Negitive values in modi review
np.where(modi['Polarity']<0)

(array([    7,     8,    15, ..., 25680, 25685, 25687], dtype=int64),)

In [16]:
#All Netural values in modi review
np.where(modi['Polarity']==0)

(array([    6,     9,    10, ..., 25682, 25683, 25686], dtype=int64),)

In [17]:
#All Positive values in rahul review
np.where(rahul['Polarity']>0)

(array([    0,     3,     5, ..., 14258, 14259, 14260], dtype=int64),)

In [18]:
#All Negitive values in rahul review
np.where(rahul['Polarity']<0)

(array([    4,     9,    10, ..., 14249, 14251, 14257], dtype=int64),)

In [19]:
#All Netural values in rahul review
np.where(rahul['Polarity']==0)

(array([    1,     2,     6, ..., 14252, 14255, 14256], dtype=int64),)

In [20]:
#Adding Label column to show the tone of all tweets in modi review
modi['Label'] = np.where(modi['Polarity']>0,'positive','negative')
modi['Label'][modi['Polarity']==0]='Neutral'

#Adding Label column to show the tone of all tweets in rahul review
rahul['Label'] = np.where(rahul['Polarity']>0,'positive','negative')
rahul['Label'][rahul['Polarity']==0]='Neutral'

C:\Users\ACER\AppData\Local\Temp\ipykernel_19464\2593688607.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  modi['Label'][modi['Polarity']==0]='Neutral'
C:\Users\ACER\AppData\Local\Temp\ipykernel_19464\2593688607.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rahul['Label'][rahul['Polarity']==0]='Neutral'


In [21]:
modi.tail()

Unnamed: 0           User  \
25683       25683     shv_indian   
25684       25684     SONUPINKOO   
25685       25685     SONUPINKOO   
25686       25686     SONUPINKOO   
25687       25687  parthajit0101   

                                                   Tweet  Polarity     Label  
25683  #1DDrive #ModiSpeaksToNews18 #LokSabhaElection...  0.000000   Neutral  
25684  I am seriously worried about future of Indian ...  0.209524  positive  
25685  If @narendramodi is so Bad then Why does @INCI... -0.700000  negative  
25686  On this 23rd March (Pakistan Day), I Urge all ...  0.000000   Neutral  
25687  @timesofindia #ArvindKejriwal, #AAP sirji pls ... -0.213542  negative

In [22]:
rahul.head()

Unnamed: 0             User  \
0           0     Sunnysweet16   
1           1    drnitinchaube   
2           2        mrvivek07   
3           3    JosephPravinP   
4           4  VandanaMegastar   

                                               Tweet  Polarity     Label  
0  Wonder why no academic or journalist asks INC ...  0.218750  positive  
1  Congrats for the change #australiavotes2019 an...  0.000000   Neutral  
2  Peopel Say “Govt Ne 70 Years Kya kiya”.\nUnse ...  0.000000   Neutral  
3  @ajaymaken @RahulGandhi And as a final touch, ...  0.320000  positive  
4  #LokSabhaElections2019 Anyone not having mass ... -0.166667  negative

In [23]:
#Removing all neutral values/tone from modi review as it is not effective in telling weather the reviews are on positive or negitive side
#originally in modi reviews there were 25688 tweets of which we are removing the netural tone tweets
#Storing all neutral tweets in neutral_modi
neutral_modi = modi[modi['Polarity']==0.0000]

remove_neutral_modi = modi['Polarity'].isin(neutral_modi['Polarity'])
modi.drop(modi[remove_neutral_modi].index,inplace=True)
print(neutral_modi.shape)
print(modi.shape)

(8207, 5)
(17481, 5)


In [24]:
#Removing all neutral values/tone from rahul review as it is not effective in telling weather the reviews are on positive or negitive side
#originally in rahul reviews there were 14261 tweets of which we are removing the netural tone tweets
#Storing all neutral tweets in neutral_rahul
neutral_rahul = rahul[rahul['Polarity']==0.0000]
remove_neutral_rahul = rahul['Polarity'].isin(neutral_rahul['Polarity'])
rahul.drop(rahul[remove_neutral_rahul].index,inplace=True)
print(neutral_rahul.shape)
print(rahul.shape)

(4894, 5)
(9367, 5)


In [25]:
print(modi.shape)
print(rahul.shape)

(17481, 5)
(9367, 5)


# Now after removing all neutral values we left with 17481 reviews for Modi which contains positive as well as negitive reviews
# And after removing all neutral values we left with 9367 reviews for Modi which contains positive as well as negitive reviews
# clearly we can see that aur data is imbalanced and now to balance data we will take in common 9000 reviews for both
# And for that we have to remove 8481 reviews from modi and 367 reviews from rahul

In [32]:
#for modi
#This will choose random reviews to delete 
np.random.seed(10)
# No of reviews to drop
remove_n = 8481
#modi.index gives us total reviews and remove_n will will give us the random 
#any 8481 reviews to delete which will be stored in drop_indices dataframe
drop_indices = np.random.choice(modi.index,remove_n,replace=False)
#here we are deleting the 8481 revieews from total reviews
df_modi = modi.drop(drop_indices)
print(df_modi.shape)

(9000, 5)


In [34]:
#same steps as above for Rahul reviews
np.random.seed(10)
remove_n = 367
drop_indices1 = np.random.choice(rahul.index,remove_n,replace=False)
df_rahul = rahul.drop(drop_indices1)
print(df_rahul.shape)

(9000, 5)


In [35]:
print(df_modi.shape)
print(df_rahul.shape)

(9000, 5)
(9000, 5)


# Prediction about indian election

In [37]:
#it will show the no of Positive and negetive  reviews in Modi
modi_count = df_modi.groupby('Label').count()
modi_count

Unnamed: 0  User  Tweet  Polarity
Label                                      
negative        2600  2600   2600      2600
positive        6400  6400   6400      6400

In [52]:
# converting the no of Pos & Neg in % where in above table Negative value 
#is on 0th index and positive value is on 1st index
neg_modi = (modi_count['Polarity'][0] / 9000) * 100
print(neg_modi)

pos_modi = (modi_count['Polarity'][1] / 9000) * 100
print(pos_modi)

28.888888888888886
71.11111111111111


In [53]:
rahul_count = df_rahul.groupby('Label').count()
rahul_count

Unnamed: 0  User  Tweet  Polarity
Label                                      
negative        3039  3039   3039      3039
positive        5961  5961   5961      5961

In [54]:
neg_rahul = (rahul_count['Polarity'][0] / 9000) * 100
print(neg_rahul)
pos_rahul = (rahul_count['Polarity'][1] / 9000) * 100
print(pos_rahul)

33.766666666666666
66.23333333333333


In [55]:
#As for prediction genrally we use Machine learninng models but this is an 
#NLP project here we have to predict the sentiment of the reviews towards modi 
#and rahul and for that we are going to plot a bar chart with the help of plotly 

In [60]:
politicians = ['Rahul','Modi']

#here we are putting the no of neg & pos reviews in a list
neg_list = [neg_rahul,neg_modi]
pos_list = [pos_rahul,pos_modi]


fig = go.Figure(
data = [
    go.Bar(name='Negative',x=politicians,y=neg_list),
    go.Bar(name='Positive',x=politicians,y=pos_list)
]
)
fig.update_layout(barmode='group')
fig.show()

# As here Modi has more positive reviews as compared to Rahul Modi wins the election